### PR TITLE
[Just In Time Messages] Send UTM params on card reader purchase links

### LINKED ISSUE #7856: [Just In Time Messages] Add UTM parameters to existing Card Reader Purchase links
## Description
In [Project Salas](https://github.com/woocommerce/woocommerce-ios/projects/51), we want to be able to understand the effectiveness of different JITM campaigns, initially those which link out to WooCommerce.com card reader purchase pages.

At the moment, it isn't practical to determine where in the app a user has tapped on a Purchase Card Reader button.

There is an industry standard for this, [called UTM, defined by the predecessor to Google Ads](https://en.wikipedia.org/wiki/UTM_parameters). UTM support is already implemented on WCcom, and added as properties on the woocommercecom_product_view and woocommercecom_page_view tracks events. Unfortunately, these are not yet added to the woocommercecom_purchased_item event, but they will be useful in funnels.

Adding UTM parameters to our links will allow us to attribute sales on WooCommerce.com, to the button in the apps the merchant tapped to get there.

It will help to ship this right at the start of the project, to establish some baseline data for the existing links, as we develop the rest of the JITM capability.

## Tasks

Add the following query parameters wherever we open links to the card reader pages. Check that they are added to the above mentioned tracks events.

- `utm_medium` – `woo_ios`
- `utm_campaign` – `feature_announcement_card`, `payments_menu_item`: identifies whether this is a Feature Announcement Card, or a settings item.
- `utm_source` – `orders_list`, `payment_method`, or `payments_menu`
- `utm_term` – {STORE_ID}
- `utm_content` – `upsell_card_readers`: identifies the specific Feature Announcement Card which the tap came from. For other display locations, it's not required

When we add other links to woocommerce.com, we will want to log similar parameters, so making helper functions for adding these parameters would be worthwhile. See pdfdoF-1yo-p2 for further context

### PR DESCRIPTION
**Pull request by joshheald**
Closes: #7856

## Description
In [Project Salas](https://github.com/woocommerce/woocommerce-ios/projects/51), we want to be able to understand the effectiveness of different JITM campaigns, initially those which link out to WooCommerce.com card reader purchase pages.

Adding UTM parameters to our links will allow us to attribute sales on WooCommerce.com, to the button in the apps the merchant tapped to get there.

## Testing instructions

Tap the CTA on the Upsell Card Reader banner (on the order list and payment methods screens) and the `Order card reader` row on the Payments menu.

Check that the following query parameters are added to the URL which is opened. 
Check that they are added to the `woocommercecom_product_view` tracks event.

- `utm_medium` – `woo_ios`
- `utm_campaign` – `feature_announcement_card`, `payments_menu_item`: identifies whether this is a Feature Announcement Card, or a settings item.
- `utm_source` – `orders_list`, `payment_method`, or `payments_menu`
- `utm_term` – {STORE_ID}
- `utm_content` – `upsell_card_readers`: identifies the specific Feature Announcement Card which the tap came from. For other display locations, it's not required

## Screenshots

CleanShot 2022-10-14 at 09 19 40@2x

CleanShot 2022-10-14 at 09 20 03@2x

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/Authentication/AuthenticatedWebViewController.swift [hunk 1]
```diff
@@ -128,6 +128,7 @@ private extension AuthenticatedWebViewController {
         guard let url = viewModel.initialURL else {
             return
         }
+        print("AuthenticatedWebViewController loadContent: \(url)")
```
Comment (toupper) [inline WooCommerce/Classes/Authentication/AuthenticatedWebViewController.swift]: Debug print here, or is it intended to be there? Perhaps we could use `DDLog`?

File: WooCommerce/Classes/Tools/WebviewHelper.swift [hunk 2]
```diff
@@ -17,6 +17,7 @@ final class WebviewHelper {
                 return
         }
 
+        print("WebviewHelper launch url \(url)")
```
Comment (toupper) [inline WooCommerce/Classes/Tools/WebviewHelper.swift]: Same here about `print` :)

File: WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/AuthenticatedWebView.swift [hunk 3]
```diff
@@ -62,6 +62,7 @@ struct AuthenticatedWebView: UIViewRepresentable {
         let parameters = ["log": username,
                           "redirect_to": url.absoluteString,
                           "authorization": "Bearer " + token]
+        print("AuthenticatedWebVew postData url \(url.absoluteString)")
```
Comment (toupper) [inline WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/AuthenticatedWebView.swift]: Same about `print` and `DDLog` :)

File: WooFoundation/WooFoundation/Utilities/WooCommerceComUTMProvider.swift [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import Foundation
+
+public struct WooCommerceComUTMProvider: UTMParametersProviding {
```
Comment (toupper) [inline WooFoundation/WooFoundation/Utilities/WooCommerceComUTMProvider.swift:3]: A neat way for making the UTM parameters generic and easily reusable! I like that 👍

**Comment by joshheald** [comment]
@toupper @pachlava N.B. I've not been able to find the events from tapping the row in settings... and I'm not sure why. 

It uses the `AuthenticatedWebViewController` from there, and that may be causing it to track differently. I've stepped through every url load/redirect on `PurchaseCardReaderWebViewViewModel.decidePolicy(for:url)`, and the query params are always where they should be, so from the app side we're doing things correctly. 

I've also followed it through with the Safari inspector connected to the webview, and the params are all there at that point too. I can't see any evidence of tracks requests from the webview, nor any that they are being blocked.

**Comment by toupper** [comment]
Nice job @joshheald! I got these links:

- Orders:

https://woocommerce.com/products/m2-card-reader/?utm_term=205113380&utm_campaign=feature_announcement_card&utm_medium=woo_ios&utm_source=order_list&utm_content=upsell_card_readers

- IPP Menu:

https://woocommerce.com/products/hardware/US?utm_source=payments_menu&utm_medium=woo_ios&utm_campaign=payments_menu_item&utm_term=205113380

- Payment Methods:

https://woocommerce.com/products/m2-card-reader/?utm_source=payment_methods&utm_term=205113380&utm_campaign=feature_announcement_card&utm_medium=woo_ios&utm_content=upsell_card_readers

These match the requirements you described above. The code looks good, I just left a few minor remarks. 👍

**Review by toupper — APPROVED** [review]

File: WooCommerce/Classes/Authentication/AuthenticatedWebViewController.swift (hunk 1 above)
Comment (joshheald) [reply WooCommerce/Classes/Authentication/AuthenticatedWebViewController.swift]: Oops. Removed. Thanks!

File: WooCommerce/Classes/Tools/WebviewHelper.swift (hunk 2 above)
Comment (joshheald) [reply WooCommerce/Classes/Tools/WebviewHelper.swift]: Oops. Removed. Thanks!

File: WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/AuthenticatedWebView.swift (hunk 3 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/AuthenticatedWebView.swift]: Oops. Removed. Thanks!

**Comment by pachlava** [comment]
Just leaving a note that I'm taking a look into it... have some troubles setting up Charles to work with iOS device

**Comment by pachlava** [comment]
I'm sorry it took me so long. I could not figure out how to set up Charles with iOS Emulator, the traffic was still encrypted after seemingly doing everything correctly. Then I went for Proxyman, and it took me just a few clicks to set it up.

The URLs look the way as you describe @joshheald:

Orders List UCR banner:

Screenshot 2022-10-14 at 13 47 30

Payment method UCR banner:

Screenshot 2022-10-14 at 13 49 47

IPP, Order card reader:

Screenshot 2022-10-14 at 13 51 49

**Pull request merged**